server/aesdsocket.c

```diff
@@ -0,0 +1,454 @@
+  if (setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) < 0) {
+    perror("setsockopt");
+    return -1;
+  }
+
+  if (create_server_address_info(address_info) != 0) {
+    syslog(LOG_ERR, "create_server_address_info");
+    return -1;
+  }
+
+  if (bind(socket_fd, (*address_info)->ai_addr, (*address_info)->ai_addrlen) ==
+      -1) {
+    perror("bind");
+    return -1;
+  }
+
+  if (listen(socket_fd, BACKLOG) == -1) {
+    perror("listen");
+    return -1;
+  }
+
+  *socket_fd_ptr = socket_fd;
+  return 0;
+}
+
+int daemonize(void) {
+  pid_t pid = fork();
+  if (pid == -1) {
+    perror("fork");
+    return -1;
+  } else if (pid != 0) {
+    // This process is the parent
+    return 1;
+  }
+
+  // Create a new session and process group
+  if (setsid() == -1) {
+    syslog(LOG_ERR, "setsid");
+    return -1;
+  }
+
+  // Set working directory to root
+  if (chdir("/") == -1) {
+    syslog(LOG_ERR, "chdir");
+    return -1;
+  }
+
+  // Note: not closing files to keep server socket open.
+
+  open("/dev/null", O_RDWR); // stdin
+  dup(0);                    // stdout
+  dup(0);                    // stderror
+
+  return 0;
+}
+
+int create_client_connection(const int server_fd, int *client_fd_ptr) {
+  struct sockaddr_in client_addr;
+  socklen_t addr_len = sizeof(struct sockaddr_in);
+
+  int client_fd = -1;
+  while (client_fd < 0) {
+    if (is_terminated) {
+      return 1;
+    }
+
+    client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &addr_len);
+    if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) {
+      struct timespec accept_delay;
+      accept_delay.tv_sec = 0;
+      accept_delay.tv_nsec = 100000000;
+      nanosleep(&accept_delay, NULL);
+      continue;
+    }
+
+    syslog(LOG_ERR, "accept: %s", strerror(errno));
+    perror("accept");
+    return -1;
+  }
+
+  char ip4_string[INET_ADDRSTRLEN];
+  inet_ntop(AF_INET, &(client_addr.sin_addr), ip4_string, INET_ADDRSTRLEN);
+  syslog(LOG_INFO, "Accepted connection from %s\n", ip4_string);
+
+  *client_fd_ptr = client_fd;
+  return 0;
+}
+
+int append_to_file(const char *file, char *buffer, const size_t buffer_len) {
+  const int fd = open(file, O_WRONLY | O_APPEND | O_CREAT,
+                      S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);
+  const ssize_t result = write(fd, buffer, buffer_len);
+  if (result == -1) {
+    perror("write");
+    return -1;
+  } else if (result != buffer_len) {
+    syslog(LOG_ERR, "partial write");
+    return -1;
+  }
+
+  if (close(fd) == -1) {
+    perror("close");
+    return -1;
+  }
+
+  return 0;
+}
+
+static int receive_and_write_data(char *file, const int client_fd) {
+  char recv_buffer[BUFFER_SIZE];
+  size_t bytes_received = 0;
+  do {
+    bytes_received = recv(client_fd, recv_buffer, sizeof(recv_buffer), 0);
+    if (bytes_received == -1) {
+      perror("recv");
+      return -1;
+    }
+
+    // Add data to end of the file
+    if (append_to_file(file, recv_buffer, bytes_received) == -1) {
```

Comment:
This design will write partial packets to the file I believe, in the case `recv` does not receive a full packet.